serix/decode.go

```diff
@@ -0,0 +1,331 @@
+	}
+	deseri := serializer.NewDeserializer(b)
+	if objectType := ts.ObjectType(); objectType != nil {
+		typeDen, objectCode, err := getTypeDenotationAndCode(objectType)
+		if err != nil {
+			return 0, errors.WithStack(err)
+		}
+		deseri.CheckTypePrefix(objectCode, typeDen, func(err error) error {
+			return errors.Wrap(err, "failed to check object type")
+		})
+	}
+	if err := api.decodeStructFields(ctx, deseri, value, valueType, opts); err != nil {
+		return 0, errors.WithStack(err)
+	}
+	return deseri.Done()
+}
+
+func (api *API) decodeStructFields(
+	ctx context.Context, deseri *serializer.Deserializer, value reflect.Value, valueType reflect.Type, opts *options,
+) error {
+	structFields, err := parseStructType(valueType)
+	if err != nil {
+		return errors.Wrapf(err, "can't parse struct type %s", valueType)
+	}
+	if len(structFields) == 0 {
+		return nil
+	}
+
+	for _, sField := range structFields {
+		fieldValue := value.Field(sField.index)
+		if sField.isEmbeddedStruct && !sField.settings.nest {
+			fieldType := sField.fType
+			if fieldType.Kind() == reflect.Ptr {
+				if fieldValue.IsNil() {
+					if sField.isUnexported {
+						return errors.Errorf(
+							"embedded field %s is a nil pointer, can't initialize because it's unexported",
+							sField.name,
+						)
+					}
+					fieldValue.Set(reflect.New(fieldType.Elem()))
+				}
+				fieldValue = fieldValue.Elem()
+				fieldType = fieldType.Elem()
+			}
+			if err := api.decodeStructFields(ctx, deseri, fieldValue, fieldType, opts); err != nil {
+				return errors.Wrapf(err, "can't deserialize embedded struct %s", sField.name)
+			}
+			continue
+		}
+		var bytesRead int
+		if sField.settings.isOptional {
+			payloadLength, err := deseri.ReadPayloadLength()
+			if err != nil {
+				return errors.Wrap(err, "can't read payload length from the deserializer")
+			}
+			if payloadLength == 0 {
+				continue
+			}
+
+			bytesRead, err = api.decode(ctx, deseri.RemainingBytes(), fieldValue, sField.settings.ts, opts)
+			if err != nil {
+				return errors.Wrapf(err, "failed to deserialize optional struct field %s", sField.name)
+			}
+			if bytesRead != int(payloadLength) {
+				return errors.Wrapf(
+					err,
+					"optional object length isn't equal to the amount of bytes read; length=%d, bytesRead=%d",
+					payloadLength, bytesRead,
+				)
+			}
+		} else {
+			bytesRead, err = api.decode(ctx, deseri.RemainingBytes(), fieldValue, sField.settings.ts, opts)
+			if err != nil {
+				return errors.Wrapf(err, "failed to deserialize struct field %s", sField.name)
+			}
+		}
+		deseri.Skip(bytesRead, func(err error) error {
+			return errors.Wrap(err, "failed to skip amount of bytes read for the struct field")
+		})
+	}
+	return nil
+}
+
+func (api *API) decodeSlice(ctx context.Context, b []byte, value reflect.Value,
+	valueType reflect.Type, ts TypeSettings, opts *options) (int, error) {
+	if valueType.AssignableTo(bytesType) {
+		lengthPrefixType, set := ts.LengthPrefixType()
```

Comment:
should probably extract code dealing with just byte slices to its own function

Reply:
Seem pretty minor to me